librosa/util/utils.py

```diff
@@ -35,80 +35,168 @@
+    (2048, 21143)
+
+    Or frame along the first axis instead of the last:
+
+    >>> frames = librosa.util.frame(y, frame_length=2048, hop_length=64, axis=0)
+    >>> frames.shape
+    (21143, 2048)
 
+    Frame a stereo signal:
+
+    >>> y, sr = librosa.load(librosa.util.example_audio_file(), mono=False)
+    >>> y.shape
+    (2, 1355168)
+    >>> frames = librosa.util.frame(y, frame_length=2048, hop_length=64)
+    (2, 2048, 21143)
+
+    Carve an STFT into fixed-length patches of 32 frames with no overlap
+
+    >>> y, sr = librosa.load(librosa.util.example_audio_file())
+    >>> S = np.abs(librosa.stft(y))
+    >>> S.shape
+    (1025, 2647)
+    >>> S_patch = librosa.util.frame(S, frame_length=32, hop_length=32)
+    >>> S_patch.shape
+    (1025, 32, 82)
+    >>> # The first patch is equivalent to the first 32 frames of S
+    >>> np.allclose(S_patch[:, :, 0], S[:, :32])
+    True
     '''
 
-    if not isinstance(y, np.ndarray):
+    if not isinstance(x, np.ndarray):
         raise ParameterError('Input must be of type numpy.ndarray, '
-                             'given type(y)={}'.format(type(y)))
+                             'given type(x)={}'.format(type(x)))
 
-    if y.ndim != 1:
-        raise ParameterError('Input must be one-dimensional, '
-                             'given y.ndim={}'.format(y.ndim))
-
-    if len(y) < frame_length:
-        raise ParameterError('Buffer is too short (n={:d})'
-                             ' for frame_length={:d}'.format(len(y), frame_length))
+    if x.shape[axis] < frame_length:
+        raise ParameterError('Input is too short (n={:d})'
+                             ' for frame_length={:d}'.format(x.shape[axis], frame_length))
 
     if hop_length < 1:
         raise ParameterError('Invalid hop_length: {:d}'.format(hop_length))
 
-    if not y.flags['C_CONTIGUOUS']:
-        raise ParameterError('Input buffer must be contiguous.')
+    n_frames = 1 + (x.shape[axis] - frame_length) // hop_length
+    strides = np.asarray(x.strides)
+
+    if axis == -1:
+        if not x.flags['F_CONTIGUOUS']:
+            raise ParameterError('Input array must be F-contiguous '
+                                 'for framing along axis={}'.format(axis))
 
-    # Compute the number of frames that will fit. The end may get truncated.
-    n_frames = 1 + int((len(y) - frame_length) / hop_length)
+        shape = list(x.shape)[:-1] + [frame_length, n_frames]
+        strides = list(strides) + [hop_length * np.prod(strides // x.itemsize) * x.itemsize]
 
-    # Vertical stride is one sample
-    # Horizontal stride is `hop_length` samples
-    y_frames = as_strided(y, shape=(frame_length, n_frames),
-                          strides=(y.itemsize, hop_length * y.itemsize))
-    return y_frames
+    elif axis == 0:
+        if not x.flags['C_CONTIGUOUS']:
+            raise ParameterError('Input array must be C-contiguous '
+                                 'for framing along axis={}'.format(axis))
```

Comment:
you can just write `'axis=0'` rather than `'axis={}'.format(axis)` in this case

Reply:
true -- it doesn't hurt though